pkg/storage/fs/ocis/tree.go

```diff
@@ -0,0 +1,282 @@
+	err = os.Remove(filepath.Join(nodePath, "parentname"))
+	if err != nil {
+		return errors.Wrap(err, "ocisfs: could not remove parent link")
+	}
+	err = os.Symlink("../"+newNode.ParentID+"/children/"+newNode.Name, filepath.Join(nodePath, "parentname"))
+	if err != nil {
+		return errors.Wrap(err, "ocisfs: could not symlink parent")
+	}
+
+	// rename child
+	err = os.Rename(
+		filepath.Join(fs.DataDirectory, "nodes", oldNode.ParentID, "children", oldNode.Name),
+		filepath.Join(fs.DataDirectory, "nodes", newNode.ParentID, "children", newNode.Name),
+	)
+	if err != nil {
+		return errors.Wrap(err, "ocisfs: could not rename symlink")
+	}
+
+	// TODO inefficient because we might update several nodes twice, only propagate unchanged nodes?
+	// collect in a list, then only stat each node once
+	// also do this in a go routine ... webdav should check the etag async
+	err = fs.Propagate(ctx, oldNode)
+	if err != nil {
+		return errors.Wrap(err, "ocisfs: Move: could not propagate old node")
+	}
+	err = fs.Propagate(ctx, newNode)
+	if err != nil {
+		return errors.Wrap(err, "ocisfs: Move: could not propagate old node")
+	}
+	return nil
+}
+
+// ChildrenPath returns the absolute path to childrens in a node
+// TODO move to node?
+func (fs *Tree) ChildrenPath(node *NodeInfo) string {
+	return filepath.Join(fs.DataDirectory, "nodes", node.ID, "children")
+}
+
+// ListFolder lists the children inside a folder
+func (fs *Tree) ListFolder(ctx context.Context, node *NodeInfo) ([]*NodeInfo, error) {
+
+	children := fs.ChildrenPath(node)
+	f, err := os.Open(children)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errtypes.NotFound(children)
+		}
+		return nil, errors.Wrap(err, "tree: error listing "+children)
+	}
+
+	names, err := f.Readdirnames(0)
+	nodes := []*NodeInfo{}
+	for i := range names {
+		link, err := os.Readlink(filepath.Join(children, names[i]))
+		if err != nil {
+			// TODO log
+			continue
+		}
+		n := &NodeInfo{
+			ParentID: node.ID,
+			ID:       filepath.Base(link),
+			Name:     names[i],
+			Exists:   true, // TODO
+		}
+
+		nodes = append(nodes, n)
+	}
+	return nodes, nil
+}
+
+// Delete deletes a node in the tree
+func (fs *Tree) Delete(ctx context.Context, node *NodeInfo) (err error) {
+	err = fs.pw.FillParentAndName(node)
+	if os.IsNotExist(err) {
+		err = nil
+		return
+	}
+	if err != nil {
+		err = errors.Wrap(err, "ocisfs: Delete: FillParentAndName error")
+		return
+	}
+
+	// remove child entry from dir
+
+	os.Remove(filepath.Join(fs.DataDirectory, "nodes", node.ParentID, "children", node.Name))
+
+	src := filepath.Join(fs.DataDirectory, "nodes", node.ID)
+	trashpath := filepath.Join(fs.DataDirectory, "trash/files", node.ID)
+	err = os.Rename(src, trashpath)
+	if err != nil {
+		return
+	}
+
+	// write a trash info ... slightly violating the freedesktop trash spec
+	t := time.Now()
+	// TODO store the original Path
+	info := []byte("[Trash Info]\nParentID=" + node.ParentID + "\nDeletionDate=" + t.Format(time.RFC3339))
```

Comment:
was that easier than using JSON ?

you also need to store the original file name for when restoring ?

Reply:
parent id and name are now stored as an extended attribute in the node